File: dotnet/src/HeatonLife.Core/MergeLife.cs
using System;
using System.Text;

namespace HeatonLife
{
    /// <summary>
    /// MergeLife cellular automaton, Heaton 2017, arXiv:1712.03019 (spec/mergelife.md).
    /// Faithful port of the reference engine (github.com/jeffheaton/mergelife),
    /// byte-identical per the upstream cross-engine conformance vectors. Do not
    /// "clean up" the numerics: the stable sort by range limit alone, the mode-padded
    /// neighbor sum, the 127/128 percent scaling, and the floor semantics are all part
    /// of the contract. State is flat row-major RGB bytes (Height*Width*3). Bit-exact tier.
    /// </summary>
    public sealed class MergeLife : IRgbFrameSource
    {
        /// <summary>"Red World", the paper's showcase genome — the Python default.</summary>
        public const string DefaultGenome = "e542-5f79-9341-f31e-6c6b-7f08-8773-7068";

        // The eight key colors: corners of the RGB cube, in genome order
        // (black, red, green, yellow, blue, purple, cyan, white).
        private static readonly int[][] ColorTable =
        {
            new[] { 0, 0, 0 },
            new[] { 255, 0, 0 },
            new[] { 0, 255, 0 },
            new[] { 255, 255, 0 },
            new[] { 0, 0, 255 },
            new[] { 255, 0, 255 },
            new[] { 0, 255, 255 },
            new[] { 255, 255, 255 },
        };

        private readonly (int Limit, double Percent, int ColorIndex)[] _rule;
        private byte[] _cells;
        private byte[] _scratch;
        private readonly int[] _avg;
        private readonly int[] _hist = new int[256];

        public string Genome { get; }
        public int Width { get; }
        public int Height { get; }
        public int Generation { get; private set; }

        /// <summary>Row-major RGB bytes, index = (y * Width + x) * 3 + channel.</summary>
        public ReadOnlySpan<byte> State => _cells;

        public MergeLife(string genome, int width, int height)
        {
            Genome = CanonicalGenome(genome);
            Width = width;
            Height = height;
            // Resolve the negative-percent flip once: the reference applies it at every
            // update (|pct| toward the *next* key color); pre-resolving is equivalent.
            var compiled = CompileRule(Genome);
            _rule = new (int, double, int)[compiled.Length];
            for (int i = 0; i < compiled.Length; i++)
            {
                var (limit, percent, colorIndex) = compiled[i];
                if (percent < 0)
                    _rule[i] = (limit, -percent, (colorIndex + 1) % ColorTable.Length);
                else
                    _rule[i] = (limit, percent, colorIndex);
            }
            _cells = new byte[width * height * 3];
            _scratch = new byte[width * height * 3];
            _avg = new int[width * height];
            SeedSoup(0); // the Python constructor's default init
        }

        /// <summary>Hex genome -> eight (range_byte, percent_byte) pairs; percent is a signed byte.</summary>
        public static (int Range, int Percent)[] ParseGenome(string genome)
        {
            string text = Normalize(genome);
            if (!IsHex32(text))
                throw new ArgumentException(
                    $"invalid MergeLife genome '{genome}' (expected 8 dash-separated groups of 4 hex digits)");
            var pairs = new (int, int)[8];
            for (int i = 0; i < 8; i++)
            {
                int range = Convert.ToInt32(text.Substring(i * 4, 2), 16);
                int pct = Convert.ToInt32(text.Substring(i * 4 + 2, 2), 16);
                if (pct >= 128) // two's complement
                    pct -= 256;
                pairs[i] = (range, pct);
            }
            return pairs;
        }

        /// <summary>Validation helper: the parse error message, or null if the genome is valid.</summary>
        public static string? GenomeError(string genome)
        {
            try
            {
                ParseGenome(genome);
            }
            catch (ArgumentException exc)
            {
                return exc.Message;
            }
            return null;
        }

        /// <summary>Normalize to the canonical lowercase dashed form.</summary>
        public static string CanonicalGenome(string genome)
        {
            ParseGenome(genome);
            string text = Normalize(genome);
            var groups = new string[8];
            for (int i = 0; i < 8; i++)
                groups[i] = text.Substring(i * 4, 4);
            return string.Join("-", groups);
        }

        /// <summary>
        /// Genome -> sub-rules (limit, percent, color_index), stably sorted by limit alone
        /// (ties keep genome order). limit = range_byte * 8, with the full-range value 2040
        /// promoted to 2048 so the top sub-rule catches every neighbor count. percent =
        /// pct/127 if positive else pct/128, giving [-1.0, 1.0]; negative percents are
        /// preserved here and resolved at application time, matching the Python compile_rule.
        /// </summary>
        public static (int Limit, double Percent, int ColorIndex)[] CompileRule(string genome)
        {
            var pairs = ParseGenome(genome);
            var entries = new (int Limit, double Percent, int ColorIndex)[8];
            for (int i = 0; i < 8; i++)
            {
                var (range, pct) = pairs[i];
                int limit = range * 8;
                if (limit == 2040)
                    limit = 2048;
                double percent = pct > 0 ? pct / 127.0 : pct / 128.0;
                entries[i] = (limit, percent, i);
            }
            // Stable sort by limit alone: ColorIndex is the genome position, so it breaks
            // ties exactly the way Python's stable sorted() does.
            Array.Sort(entries, (a, b) =>
                a.Limit != b.Limit ? a.Limit.CompareTo(b.Limit) : a.ColorIndex.CompareTo(b.ColorIndex));
            return entries;
        }

        /// <summary>A random genome from PCG32 (UI convenience; not part of the upstream contract).</summary>
        public static string RandomGenome(uint seed)
        {
            var rng = new Pcg32(seed);
            var sb = new StringBuilder();
            for (int i = 0; i < 8; i++)
            {
                if (i > 0)
                    sb.Append('-');
                uint range = rng.NextU32() & 0xFF;
                uint pct = rng.NextU32() & 0xFF;
                sb.Append(range.ToString("x2")).Append(pct.ToString("x2"));
            }
            return sb.ToString();
        }

        private static string Normalize(string genome) =>
            genome.Trim().ToLowerInvariant().Replace("-", "");

        private static bool IsHex32(string text)
        {
            if (text.Length != 32)
                return false;
            foreach (char c in text)
                if (!((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f')))
                    return false;
            return true;
        }

        /// <summary>Soup init per spec: one PCG32 draw per byte (draw &amp; 0xFF), row-major, RGB innermost.</summary>
        public void SeedSoup(ulong seed)
        {
            var rng = new Pcg32(seed);
            for (int i = 0; i < _cells.Length; i++)
                _cells[i] = (byte)(rng.NextU32() & 0xFF);
            Generation = 0;
        }

        /// <summary>Load an explicit RGB lattice (row-major, Height*Width*3 bytes).</summary>
        public void SetState(ReadOnlySpan<byte> cells)
        {
            if (cells.Length != _cells.Length)
                throw new ArgumentException($"expected {_cells.Length} bytes, got {cells.Length}");
            cells.CopyTo(_cells);
            Generation = 0;
        }

        /// <summary>Restore a saved state at a given generation (catalog loads).</summary>
        public void SetState(ReadOnlySpan<byte> cells, int generation)
        {
            SetState(cells);
            Generation = generation;
        }

        /// <summary>Export the RGB lattice as PNG, each cell a scale×scale block (spec/png-io.md).</summary>
        public byte[] ToPng(int scale = 1) => PngGrid.EncodeRgb(_cells, Width, Height, scale);

        /// <summary>Paint one cell's RGB in place; does not reset the generation.</summary>
        public void SetCell(int x, int y, byte r, byte g, byte b)
        {
            if (x < 0 || x >= Width || y < 0 || y >= Height)
                throw new ArgumentOutOfRangeException($"cell ({x}, {y}) outside {Width}x{Height}");
            int index = (y * Width + x) * 3;
            _cells[index] = r;
            _cells[index + 1] = g;
            _cells[index + 2] = b;
        }

        public void Step(int n = 1)
        {
            for (int s = 0; s < n; s++)
                StepOnce();
            Generation += n;
        }

        /// <summary>Frame per spec/render.md: the state itself, raw RGB bytes.</summary>
        public void WriteFrame(byte[] rgb)
        {
            if (rgb.Length != _cells.Length)
                throw new ArgumentException($"expected {_cells.Length} bytes, got {rgb.Length}");
            Array.Copy(_cells, rgb, _cells.Length);
        }

        /// <summary>Allocating convenience for <see cref="WriteFrame"/>.</summary>
        public byte[] Frame()
        {
            var rgb = new byte[_cells.Length];
            WriteFrame(rgb);
            return rgb;
        }

        private void StepOnce()
        {
            int w = Width, h = Height, cells = w * h;

            // Merged grid: integer floor((r+g+b)/3), and its histogram for the mode.
            Array.Clear(_hist, 0, 256);
            for (int i = 0; i < cells; i++)
            {
                int avg = (_cells[i * 3] + _cells[i * 3 + 1] + _cells[i * 3 + 2]) / 3;
                _avg[i] = avg;
                _hist[avg]++;
            }
            // Grid mode; first maximum wins, i.e. ties break toward the lowest value,
            // same as the reference's bincount().argmax().
            int mode = 0, best = -1;
            for (int v = 0; v < 256; v++)
            {
                if (_hist[v] > best)
                {
                    best = _hist[v];
                    mode = v;
                }
            }

            for (int y = 0; y < h; y++)
            {
                for (int x = 0; x < w; x++)
                {
                    // 3x3 neighbor sum of the merged grid (center excluded),
                    // out-of-bounds cells padded with the mode.
                    int cnt = 0;
                    for (int dy = -1; dy <= 1; dy++)
                        for (int dx = -1; dx <= 1; dx++)
                        {
                            if (dy == 0 && dx == 0)
                                continue;
                            int ny = y + dy, nx = x + dx;
                            cnt += (ny < 0 || ny >= h || nx < 0 || nx >= w)
                                ? mode
                                : _avg[ny * w + nx];
                        }

                    // First sub-rule (in sorted order) with cnt < limit wins;
                    // unmatched cells keep their value.
                    int baseIndex = (y * w + x) * 3;
                    bool matched = false;
                    foreach (var (limit, percent, colorIndex) in _rule)
                    {
                        if (cnt >= limit)
                            continue;
                        int[] color = ColorTable[colorIndex];
                        for (int c = 0; c < 3; c++)
                        {
                            int value = _cells[baseIndex + c];
                            int moved = value + (int)Math.Floor((color[c] - value) * percent);
                            _scratch[baseIndex + c] = (byte)moved;
                        }
                        matched = true;
                        break;
                    }
                    if (!matched)
                    {
                        _scratch[baseIndex] = _cells[baseIndex];
                        _scratch[baseIndex + 1] = _cells[baseIndex + 1];
                        _scratch[baseIndex + 2] = _cells[baseIndex + 2];
                    }
                }
            }
            (_cells, _scratch) = (_scratch, _cells);
        }
    }
}

File: python/src/heaton_life/ca/mergelife.py
"""MergeLife cellular automaton (Heaton 2017, arXiv:1712.03019) — spec/mergelife.md.

Faithful port of the reference engine (github.com/jeffheaton/mergelife,
python/mergelife-lib), byte-identical per the upstream cross-engine conformance
vectors replayed in tests/test_mergelife.py. Do not "clean up" the numerics:
the stable sort by range limit alone, the mode-padded neighbor convolution, the
127/128 percent scaling, and the floor semantics are all part of the contract.
"""

from __future__ import annotations

import dataclasses
import re

import numpy as np
from numpy.typing import NDArray

from heaton_life.core.params import Params
from heaton_life.core.rng import Pcg32
from heaton_life.init.patterns import clear

# The eight key colors: corners of the RGB cube, in rule order.
COLOR_TABLE = np.array(
    [
        [0, 0, 0],  # black
        [255, 0, 0],  # red
        [0, 255, 0],  # green
        [255, 255, 0],  # yellow
        [0, 0, 255],  # blue
        [255, 0, 255],  # purple
        [0, 255, 255],  # cyan
        [255, 255, 255],  # white
    ],
    dtype=np.int64,
)

_RULE = re.compile(r"^[0-9a-f]{32}$")


def parse_rule(rule: str) -> list[tuple[int, int]]:
    """Hex rule -> eight (range_byte, percent_byte) pairs; percent is a signed byte."""
    text = rule.strip().lower().replace("-", "")
    if not _RULE.match(text):
        raise ValueError(
            f"invalid MergeLife rule {rule!r} (expected 8 dash-separated groups of 4 hex digits)"
        )
    pairs: list[tuple[int, int]] = []
    for i in range(8):
        rng = int(text[i * 4 : i * 4 + 2], 16)
        pct = int(text[i * 4 + 2 : i * 4 + 4], 16)
        if pct >= 128:  # two's complement
            pct -= 256
        pairs.append((rng, pct))
    return pairs


def parse_rule_error(rule: str) -> str | None:
    """Validation helper: the parse error message, or None if the rule is valid."""
    try:
        parse_rule(rule)
    except ValueError as exc:
        return str(exc)
    return None


def canonical_rule(rule: str) -> str:
    """Normalize to the canonical lowercase dashed form."""
    parse_rule(rule)
    text = rule.strip().lower().replace("-", "")
    return "-".join(text[i * 4 : i * 4 + 4] for i in range(8))


def compile_rule(rule: str) -> list[tuple[int, float, int]]:
    """Rule -> sub-rules (limit, percent, color_index), stably sorted by limit alone.

    limit = range_byte * 8, with the full-range value 2040 promoted to 2048 so the
    top sub-rule catches every neighbor count. percent = pct/127 if positive else
    pct/128, giving [-1.0, 1.0].
    """
    entries: list[tuple[int, float, int]] = []
    for index, (rng, pct) in enumerate(parse_rule(rule)):
        limit = rng * 8
        if limit == 2040:
            limit = 2048
        percent = pct / 127.0 if pct > 0 else pct / 128.0
        entries.append((limit, percent, index))
    return sorted(entries, key=lambda e: e[0])  # stable: ties keep rule order


COLOR_NAMES = ("Black", "Red", "Green", "Yellow", "Blue", "Purple", "Cyan", "White")


@dataclasses.dataclass(frozen=True)
class DecodedSubRule:
    """One row of the rule-lab table — spec/mergelife.md "Decoded rule table"."""

    limit: int
    range_low: int
    range_high: int
    percent: float
    color_index: int
    color_name: str
    target_index: int
    target_name: str
    target_rgb: tuple[int, int, int]
    range_byte: int
    percent_byte: int


def decode_rule(rule: str) -> list[DecodedSubRule]:
    """The 8 sub-rules as display-ready rows, in compiled (sorted) order.

    Carries the raw octets alongside the compiled values: HeatonCA re-derives
    octet 1 from the promoted limit (2048/8 = 0x100) — the raw byte is 0xff.
    """
    raw = parse_rule(rule)
    rows: list[DecodedSubRule] = []
    low = 0
    for limit, percent, index in compile_rule(rule):
        range_byte, percent_byte = raw[index]
        target = (index + 1) % len(COLOR_NAMES) if percent < 0 else index
        r, g, b = (int(c) for c in COLOR_TABLE[target])
        rows.append(
            DecodedSubRule(
                limit=limit,
                range_low=low,
                range_high=limit - 1,
                percent=percent,
                color_index=index,
                color_name=COLOR_NAMES[index],
                target_index=target,
                target_name=COLOR_NAMES[target],
                target_rgb=(r, g, b),
                range_byte=range_byte,
                percent_byte=percent_byte,
            )
        )
        low = limit
    return rows


def random_rule(seed: int) -> str:
    """A random rule from PCG32 (UI convenience; not part of the upstream contract)."""
    rng = Pcg32(seed)
    groups = []
    for _ in range(8):
        range_byte = rng.next_u32() & 0xFF
        pct_byte = rng.next_u32() & 0xFF
        groups.append(f"{range_byte:02x}{pct_byte:02x}")
    return "-".join(groups)


@dataclasses.dataclass(frozen=True)
class MergeLifeParams(Params):
    rule: str = dataclasses.field(
        default="e542-5f79-9341-f31e-6c6b-7f08-8773-7068",  # "Red World" (paper)
        metadata={"label": "Rule"},
    )
    width: int = dataclasses.field(default=128, metadata={"min": 8, "max": 1024})
    height: int = dataclasses.field(default=128, metadata={"min": 8, "max": 1024})
    init: str = dataclasses.field(default="soup", metadata={"choices": ["soup"]})
    seed: int = dataclasses.field(
        default=0, metadata={"min": 0, "max": 4294967295, "role": "seed"}
    )


class MergeLife:
    """MergeLife CA. State is uint8 RGB, shape (height, width, 3), row-major."""

    def __init__(
        self,
        rule: str = "e542-5f79-9341-f31e-6c6b-7f08-8773-7068",
        *,
        size: tuple[int, int] = (128, 128),
        init: str | NDArray[np.uint8] = "soup",
        seed: int = 0,
    ) -> None:
        width, height = size
        self._initial: NDArray[np.uint8] | None = None
        if isinstance(init, np.ndarray):
            if init.shape != (height, width, 3):
                raise ValueError(
                    f"init array shape {init.shape} does not match (h={height}, w={width}, 3)"
                )
            self._initial = init.astype(np.uint8)
            init_name = "array"
        else:
            init_name = init
        self.params = MergeLifeParams(
            rule=canonical_rule(rule),
            width=width,
            height=height,
            init=init_name,
            seed=seed,
        )
        self._rule = compile_rule(self.params.rule)
        self._generation = 0
        self._state: NDArray[np.uint8]
        self.reset()

    @classmethod
    def from_params(cls, params: MergeLifeParams) -> MergeLife:
        if params.init == "array":
            raise ValueError("params with init='array' need the array: MergeLife(init=...)")
        return cls(
            params.rule,
            size=(params.width, params.height),
            init=params.init,
            seed=params.seed,
        )

    def reset(self, seed: int | None = None) -> None:
        if seed is not None:
            self.params = self.params.replace(seed=seed)
        p = self.params
        if p.init == "array":
            assert self._initial is not None
            self._state = self._initial.copy()
        elif p.init == "soup":
            rng = Pcg32(p.seed)
            flat = np.fromiter(
                (rng.next_u32() & 0xFF for _ in range(p.width * p.height * 3)),
                dtype=np.uint8,
                count=p.width * p.height * 3,
            )
            self._state = flat.reshape(p.height, p.width, 3)
        else:
            raise ValueError(f"unknown init strategy: {p.init!r}")
        self._generation = 0

    def step(self, n: int = 1) -> None:
        for _ in range(n):
            self._state = _update(self._state, self._rule)
        self._generation += n

    @property
    def state(self) -> NDArray[np.uint8]:
        return self._state

    @property
    def generation(self) -> int:
        return self._generation

    def clear_rect(self, x0: int, y0: int, x1: int, y1: int) -> None:
        """spec/patterns.md "Clear": inclusive rectangle to this family's blank.

        MergeLife's blank is black (0, 0, 0); the trailing RGB axis broadcasts. The generation is untouched, exactly like stamping. The C# port has
        carried this per family as ``ClearRect`` all along; the Python library had
        no equivalent at all, even though the spec lists Clear alongside Extract
        and Stamp as normative.
        """
        clear(self._state, x0, y0, x1, y1, 0)

    def frame(self) -> NDArray[np.uint8]:
        return self._state.copy()


def _update(
    prev: NDArray[np.uint8], rule: list[tuple[int, float, int]]
) -> NDArray[np.uint8]:
    """One MergeLife step, exactly per the reference engine."""
    height, width = prev.shape[:2]

    # Merged grid: integer floor((r+g+b)/3).
    avg = prev.astype(np.int64).sum(axis=2) // 3
    # Grid mode of the merged values; bincount().argmax() breaks ties toward the
    # lowest value, same as the reference.
    pad_val = int(np.bincount(avg.ravel()).argmax())
    # 3x3 neighbor sum (center excluded), boundary padded with the mode.
    padded = np.full((height + 2, width + 2), pad_val, dtype=np.int64)
    padded[1:-1, 1:-1] = avg
    cnt = (
        padded[:-2, :-2]
        + padded[:-2, 1:-1]
        + padded[:-2, 2:]
        + padded[1:-1, :-2]
        + padded[1:-1, 2:]
        + padded[2:, :-2]
        + padded[2:, 1:-1]
        + padded[2:, 2:]
    )

    # Unmatched cells keep their value.
    new = prev.copy()
    remaining = np.ones((height, width), dtype=bool)
    for limit, percent, color_index in rule:
        mask = (cnt < limit) & remaining
        if not mask.any():
            continue
        remaining &= ~mask
        if percent < 0:
            percent = -percent
            color_index = (color_index + 1) % len(COLOR_TABLE)
        sel = prev[mask].astype(np.int64)
        moved = sel + np.floor((COLOR_TABLE[color_index] - sel) * percent)
        new[mask] = moved.astype(np.uint8)
    return new
